### PR TITLE
auth: add basic tests

### PR DESCRIPTION
**Pull request by gyuho**
/cc @mitake

### CODE REVIEW

**Comment by mitake** [comment]
@gyuho lgtm, thanks

File: auth/store_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/coreos/etcd/auth/authpb"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/mvcc/backend"
+)
+
+func TestUserAddAndAuthenticate(t *testing.T) {
```
Comment (xiang90) [inline auth/store_test.go]: let's test exact one functionality in one test. we should separate this to two tests.

File: auth/store_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		os.Remove(tPath)
+	}()
+
+	as := NewAuthStore(b)
+	ua := &pb.AuthUserAddRequest{Name: "foo", Password: "bar"}
+	_, err := as.UserAdd(ua)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserAdd(ua)
```
Comment (xiang90) [inline auth/store_test.go]: to test user add, we need to test two cases: add a non-existing user and add an existing user.

File: auth/store_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		if err != ErrUserAlreadyExist {
+			t.Fatalf("expected %v, got %v", ErrUserAlreadyExist, err)
+		}
+	} else {
+		t.Fatalf("expected %v, got %v", ErrUserAlreadyExist, err)
+	}
+
+	_, err = as.Authenticate("foo", "bar")
```
Comment (xiang90) [inline auth/store_test.go]: for authenticate we should test a few cases:

1. auth a non-existing user
2. auth an existing user but with wrong password
3. auth an existing user with correct password

File: auth/store_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	_, err := as.UserAdd(ua)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ud := &pb.AuthUserDeleteRequest{Name: "foo"}
+	_, err = as.UserDelete(ud)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserDelete(ud)
```
Comment (xiang90) [inline auth/store_test.go]: we should also test deleting a non-existing user.

File: auth/store_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		if err != ErrUserNotFound {
+			t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+		}
+	} else {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+}
+
+func TestUserChangePassword(t *testing.T) {
```
Comment (xiang90) [inline auth/store_test.go]: to test change password, we need to test change a non-existing user; change existing user with wrong password

File: auth/store_test.go (hunk 1 above)
Comment (gyuho) [inline auth/store_test.go]: Ok will address all in the next commit. Thanks.

File: auth/store_test.go (hunk 5 above)
Comment (gyuho) [inline auth/store_test.go]: `change existing user with wrong password` cannot be tested with current implementation. It just overwrites with new password. It doesn't check if it provides the correct previous password.

**Comment by gyuho** [comment]
@xiang90 All addressed except `ChangePassword`. PTAL.

File: auth/store_test.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	}()
+
+	as := NewAuthStore(b)
+	ua := &pb.AuthUserAddRequest{Name: "foo"}
+	_, err := as.UserAdd(ua) // add a non-existing user
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserAdd(ua) // add an existing user
+	if err != nil {
```
Comment (xiang90) [inline auth/store_test.go]: if err!=nil && err != ErrUserAlreadyExist

File: auth/store_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+		b.Close()
+		os.Remove(tPath)
+	}()
+
+	as := NewAuthStore(b)
+
+	// auth a non-existing user
+	_, err := as.Authenticate("foo", "bar")
+	if err != nil {
+		if !strings.Contains(err.Error(), "auth: authentication failed, invalid user ID or password") {
```
Comment (xiang90) [inline auth/store_test.go]: we probably should define this error?

File: auth/store_test.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+			t.Fatalf("expected 'auth: authentication failed, invalid user ID or password' error, got %v", err)
+		}
+	} else {
+		t.Fatalf("expected 'invalid password...' error, got %v", err)
+	}
+
+	// auth an existing user with correct password
+	ua := &pb.AuthUserAddRequest{Name: "foo", Password: "bar"}
```
Comment (xiang90) [inline auth/store_test.go]: move the user add to the top? before we start doing any test.

File: auth/store_test.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	// delete an existing user
+	ud := &pb.AuthUserDeleteRequest{Name: "foo"}
+	_, err = as.UserDelete(ud)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// delete a non-existing user
+	_, err = as.UserDelete(ud)
+	if err != nil {
```
Comment (xiang90) [inline auth/store_test.go]:
```go
if err == nil {
    ....
}
if err != ErrUserNotFound {
...
}
```

There are a few other places we should change to this. we should minimize the indentation.

File: auth/store_test.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+		if err != ErrUserNotFound {
+			t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+		}
+	} else {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+}
+
+func TestRoleAddGrant(t *testing.T) {
```
Comment (xiang90) [inline auth/store_test.go]: This test tests too many stuff. we should not test permission grant in this test.

File: auth/store_test.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	// grants a permission to a role
+	_, err = as.RoleGrant(&pb.AuthRoleGrantRequest{Name: "test-role", Perm: &authpb.Permission{PermType: authpb.READ}})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// grants a role to the user
+	_, err = as.UserAdd(&pb.AuthUserAddRequest{Name: "foo"})
```
Comment (xiang90) [inline auth/store_test.go]: we also need to test grant a role to a non-existing user.

Comment (gyuho) [inline auth/store_test.go]: sure, will address all in the next commit. thanks.

**Comment by gyuho** [comment]
@xiang90 All addressed (reduce indentation, use defined error types, split tests). PTAL.

File: auth/store_test.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+	}()
+
+	as := NewAuthStore(b)
+	ua := &pb.AuthUserAddRequest{Name: "foo"}
+	_, err := as.UserAdd(ua) // add a non-existing user
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserAdd(ua) // add an existing user
+	if (err != nil && err != ErrUserAlreadyExist) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]:
```go
if err == nil {
    fatalf
}
if err != nil && err != ErrUserAlreadyExist {
    ....
}
```

File: auth/store_test.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+	// auth an existing user with correct password
+	_, err = as.Authenticate("foo", "bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// auth an existing user but with wrong password
+	_, err = as.Authenticate("foo", "")
+	if (err != nil && err != ErrAuthFailed) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]: same here

File: auth/store_test.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+	ua := &pb.AuthUserAddRequest{Name: "foo", Password: "bar"}
+	_, err := as.UserAdd(ua)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// auth a non-existing user
+	_, err = as.Authenticate("foo-test", "bar")
+	if (err != nil && err != ErrAuthFailed) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]: same here

File: auth/store_test.go [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+	// delete an existing user
+	ud := &pb.AuthUserDeleteRequest{Name: "foo"}
+	_, err = as.UserDelete(ud)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// delete a non-existing user
+	_, err = as.UserDelete(ud)
+	if (err != nil && err != ErrUserNotFound) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]: same here.

File: auth/store_test.go [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+		t.Fatal(err)
+	}
+	_, err = as.Authenticate("foo", "bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// change a non-existing user
+	_, err = as.UserChangePassword(&pb.AuthUserChangePasswordRequest{Name: "foo-test", Password: "bar"})
+	if (err != nil && err != ErrUserNotFound) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]: same here.

File: auth/store_test.go [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+	}()
+
+	as := NewAuthStore(b)
+
+	_, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "foo"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// grants a permission to a non-existing role
```
Comment (xiang90) [inline auth/store_test.go]: remove this test. we should only test role grant, not permission grant.

File: auth/store_test.go [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	// adds a new role
+	_, err = as.RoleAdd(&pb.AuthRoleAddRequest{Name: "role-test"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// grants a permission to a role
+	_, err = as.RoleGrant(&pb.AuthRoleGrantRequest{Name: "role-test", Perm: &authpb.Permission{PermType: authpb.READ}})
```
Comment (xiang90) [inline auth/store_test.go]: remove this too.

File: auth/store_test.go [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+	// grants a role to the user
+	_, err = as.UserGrant(&pb.AuthUserGrantRequest{User: "foo", Role: "role-test"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// grants a role to a non-existing user
+	_, err = as.UserGrant(&pb.AuthUserGrantRequest{User: "foo-test", Role: "role-test"})
+	if (err != nil && err != ErrUserNotFound) || err == nil {
```
Comment (xiang90) [inline auth/store_test.go]:
```go
if err == nil {
    fatalf
}
if err != nil && err != ErrUserNotFound {
    ....
}
```

**Comment by gyuho** [comment]
@xiang90 PTAL. Cleaned up error checking and fix `user  grant` test.

File: auth/store_test.go [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+	ua := &pb.AuthUserAddRequest{Name: "foo"}
+	_, err := as.UserAdd(ua) // add a non-existing user
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = as.UserAdd(ua) // add an existing user
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrUserAlreadyExist, err)
+	}
+	if err != nil && err != ErrUserAlreadyExist {
```
Comment (xiang90) [inline auth/store_test.go]: remove err!=nil

File: auth/store_test.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// auth a non-existing user
+	_, err = as.Authenticate("foo-test", "bar")
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrAuthFailed, err)
+	}
+	if err != nil && err != ErrAuthFailed {
```
Comment (xiang90) [inline auth/store_test.go]: remove err!=nil

File: auth/store_test.go [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// auth an existing user but with wrong password
+	_, err = as.Authenticate("foo", "")
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrAuthFailed, err)
+	}
+	if err != nil && err != ErrAuthFailed {
```
Comment (xiang90) [inline auth/store_test.go]: same here.

File: auth/store_test.go [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// delete a non-existing user
+	_, err = as.UserDelete(ud)
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+	if err != nil && err != ErrUserNotFound {
```
Comment (xiang90) [inline auth/store_test.go]: and here.

File: auth/store_test.go [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// change a non-existing user
+	_, err = as.UserChangePassword(&pb.AuthUserChangePasswordRequest{Name: "foo-test", Password: "bar"})
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+	if err != nil && err != ErrUserNotFound {
```
Comment (xiang90) [inline auth/store_test.go]: and here.

File: auth/store_test.go [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// grants a role to a non-existing user
+	_, err = as.UserGrant(&pb.AuthUserGrantRequest{User: "foo-test", Role: "role-test"})
+	if err == nil {
+		t.Fatalf("expected %v, got %v", ErrUserNotFound, err)
+	}
+	if err != nil && err != ErrUserNotFound {
```
Comment (xiang90) [inline auth/store_test.go]: and here.

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
removed unnecessary `nil` checking. merging in greens. thanks.

**Pull request merged**